Add Type 1 transform for gridded arrays

tyler-a-cox opened the pull request:
For cases when arrays are gridded, Type 1 transforms can make `fftvis` substantially faster, especially when the maximum extent of the array is large. This PR implements methods to detect whether an input array is gridded and automatically triggers the use of Type 1 if the array is gridded.

tyler-a-cox committed:
move initial beam interpolation to the wrapper script

tyler-a-cox committed:
new utility functions for checking if array is griddable

tyler-a-cox committed:
add type 1 transform to cpu_nufft

tyler-a-cox committed:
make function that checks gridding accessible

tyler-a-cox committed:
add type 1 simulation branch

tyler-a-cox committed:
add missing import

tyler-a-cox committed:
add utility tests

tyler-a-cox committed:
check the correct axis

tyler-a-cox committed:
remove redundant code

tyler-a-cox committed:
fix hex antenna positions for gridding test

steven-murray commented on src/fftvis/core/utils.py:
```diff
@@ -206,3 +223,134 @@ def inplace_rot_base(rot, b):
+    return False, f
+
+
+def check_antpos_griddability(
```
I suggest we add an argument like "rotation_matrix" to this, so the user can supply their own transform in the case that it's not a rectangular/hexagonal layout.

tyler-a-cox replied:
Added an argument that allows the user to pass in their own rotation matrix.

steven-murray replied:
Sweet, thanks!

steven-murray commented on src/fftvis/cpu/cpu_nufft.py:
```diff
@@ -108,3 +108,52 @@ def cpu_nufft3d(
+        Weights of sources (typically beam-weighted fluxes).
+    n_modes : int
+        Number of modes in the transform.
+    index : np.ndarray
```
If this is to be a public-facing API, then I suggest documenting this argument better (e.g. say that it needs to be a 2D array,a nd what the shape corresponds to).

tyler-a-cox replied:
Added slightly more description in the docstring.

steven-murray commented on src/fftvis/cpu/cpu_simulate.py:
```diff
@@ -391,12 +437,14 @@ def _evaluate_vis_chunk(
 
                 # Rotate source coordinates with rotation matrix.
-                utils.inplace_rot(rotation_matrix, topo)
+                inplace_rot(rotation_matrix, topo)
```
Does the math not work out to be able to do the rotation on the equatorial coordinates, so we only have to do it once?

tyler-a-cox replied:
I think there are some cases when you can avoid doing this. Certainly, if the rotation matrix is the identity, we could just skip this, but I think if the rotation matrix requires shearing like with the hex array, I don't think we can pre-rotate, but I may be wrong. I'll have a think about this, but in the meantime, I'll add a check that passes on this if the rotation matrix is the identity.

steven-murray replied:
I think we want to stick away from "rotation matrix" terminology because rotation matrices don't shear. It's not obvious to me that transforming baselines with a shear is able to reproduce the same visibilities after inversely transforming the source positions, but it seems like it does! I'd love to be convinced of this :-)

tyler-a-cox replied:
Yeah, good point! I've added a new parameter to `_evaluate_vis_chunk` that carries the basis transformation matrix. I've added a series of new tests that compares visibilities calculated with the Type 1 transform against matvis for gridded matrices that have been skewed, rotated, etc.

steven-murray commented on tests/test_core_utils.py:
```diff
@@ -231,3 +231,58 @@ def mock_import_error(name, *args, **kwargs):
         # Restore original import
         monkeypatch.undo()
+
+def test_check_array_griddability():
```
I suggest making this test a class: `TestCheckArrayGriddability` and then make each subtest a new test function (or maybe use parametrization). That will help with test failure identification.

tyler-a-cox replied:
Reorganized this using parametrization.

steven-murray replied:
I didn't see the updated parameterized test, so maybe you forgot to push that?

tyler-a-cox replied:
You're right, I'd thought I'd pushed that earlier. Should be updated now!

steven-murray reviewed (commented):
Hey Tyler, this is super cool, thanks! I have a few broader questions/comments:

1. I would move the new `cpu_utils.py` stuff into its own module (like `antenna_gridding.py` or something nicer)
2. I don't know why we have `cpu/cpu_nufft.py` etc instead of just `cpu/nufft.py` which is much nicer.
3. How does the antenna rotation work? Rigid-Rotating a hex array should not be able to shear the antennas, right?
4. I would like to see a test comparing a gridded to non-gridded calculation for the same ant positions and sky model. This would mean we also need an option to tell the simulator to NOT try gridding even if it could grid.
5. I also think that we should let the user tell the simulate function that yes, indeed, I want to use gridding, and I know just how to grid it. Probably by passing a rotation matrix and argument like "griddable = True" (or maybe just the rotation matrix). That way we can skip the checks and just assume the user knows what they're doing. Helps with speed a little bit.

tyler-a-cox committed:
reorganize tests and gridding code

steven-murray commented on src/fftvis/cpu/cpu_simulate.py:
```diff
@@ -110,7 +111,9 @@ def simulate(
             "CoordinateRotationAstropy", "CoordinateRotationERFA"
         ] = "CoordinateRotationERFA",
         coord_method_params: dict | None = None,
+        skew_to_cartesian: np.ndarray | None = None,
```
Did we add docstrings for these into the base class?

tyler-a-cox replied:
Removed this in favor of a more general gridding function that finds a set of basis vectors for the basis transformation.

steven-murray commented on src/fftvis/wrapper.py:
```diff
@@ -174,6 +176,15 @@ def simulate_vis(
+        to a Cartesian grid that is compatible with a more efficient NUFFT algorithm. By default fftvis checks
+        if the antenna positions are on a regular or hexagonal grid, but other grid types are not automatically
+        supported. The matrix should be of shape (3, 3).
+    force_use_type3: bool, default = False
+        Whether to force the use of type 3 NUFFT. If False, type 3 will only be used
+        if the array can be defined in a regular grid.
```
Isn't this the wrong way around? Type 3 will be used if the array *cannot* be defined on a regular grid?

tyler-a-cox replied:
Yeah, that's right. I'll change that. Thanks for the catch.

steven-murray commented on tests/test_cpu_simulate.py:
```diff
@@ -148,11 +158,12 @@ def test_simulate(
 
     # Check shape of result when no baselines are specified
-    nbls = (len(ants) * (len(ants) - 1)) // 2 + 1
+    nbls = int((len(ants) * (len(ants) - 1)) / 2)
```
Why did this change?

tyler-a-cox replied:
It was just differences in number of baselines returned due to me moving from a random array to a regular array. I've changed this back and separated the gridded tests.

tyler-a-cox committed:
generalize antenna gridding

tyler-a-cox committed:
fix transposition error and remove old utility functions

tyler-a-cox committed:
add upsampfac parameter to increase accuracy of precision=1 type 3 transform

tyler-a-cox committed:
add documentation of the upsampfac parameter

tyler-a-cox committed:
rename cpu/gpu modules

tyler-a-cox commented:
> Hey Tyler, this is super cool, thanks! I have a few broader questions/comments:
> 
> 1. I would move the new `cpu_utils.py` stuff into its own module (like `antenna_gridding.py` or something nicer)
> 2. I don't know why we have `cpu/cpu_nufft.py` etc instead of just `cpu/nufft.py` which is much nicer.
> 3. How does the antenna rotation work? Rigid-Rotating a hex array should not be able to shear the antennas, right?
> 4. I would like to see a test comparing a gridded to non-gridded calculation for the same ant positions and sky model. This would mean we also need an option to tell the simulator to NOT try gridding even if it could grid.
> 5. I also think that we should let the user tell the simulate function that yes, indeed, I want to use gridding, and I know just how to grid it. Probably by passing a rotation matrix and argument like "griddable = True" (or maybe just the rotation matrix). That way we can skip the checks and just assume the user knows what they're doing. Helps with speed a little bit.

Hey Steven, thanks for taking a look!

1. All the gridding code has now been moved to the `antenna_gridding.py` module.
2. I've renamed these modules to `nufft.py` rather than `cpu_nufft.py`
3. Yeah, you're right, I think the "rotation" name in the code is a bit misleading. In practice, the code is much more general. As long as you can define some basis that can map the physical antenna positions onto integer grid points, then you should be able to use the Type 1 in place of the Type 3. It only really requires that you can apply the inverse of that matrix to the source coordinates and use the antenna positions projected in their new coordinates as indexing arrays. It also requires being a little careful about getting the scaling of the coordinates right, but it does seem to work pretty well.
4. Yeah, I do think it's a good idea to let users have the option to force `fftvis` to use the type 3 ever when the type 1 can be used. I've added a new keyword argument `force_use_type3` which requires `fftvis` to use the type 3. I've also added a parameterized test function that compares the Type 1 and Type 3 to matvis for a hex array and a rectangularly gridded array. I've also applied shears, rotations, and removed antennas to test the gridding function and make sure that the gridding is being done sensibly.
5. I decided to go a slightly different route on this point and just make the `check_antpos_griddability` better at detecting a range of gridded arrays. The code now looks for the two shortest, non-collinear baselines in the array to use as basis functions that are used for gridding.

tyler-a-cox commented:
I've also caught a potential issue using `precision=1`. I noticed that the Type 1 actually gave results that more closely match the Type 3 when using single precision. After digging into the `finufft` code, it looks like the Type 3 uses an upsampling factor of 1.25 for single precision, while the Type 1 uses an upsampling factor of 2.0. I decided to add another parameter that sets the upsampling factor (default=2.0) to maintain accuracy at single precision.

tyler-a-cox committed:
revert to {cpu/gpu}_simulate

steven-murray commented on src/fftvis/core/antenna_gridding.py:
```diff
@@ -0,0 +1,194 @@
+    antpos : dict
+        Dictionary of antenna positions in the form {ant_index: np.array([x,y,z])}.
+    tol : float
+        Tolerance for checking if values are close to integers.
```
This is ultimately in meters, right? Or, in the units of whatever antpos is in? We should say so.

tyler-a-cox replied:
Changed this.

steven-murray commented on src/fftvis/core/simulate.py:
```diff
@@ -37,6 +37,7 @@ def simulate(
         precision: int = 2,
         polarized: bool = False,
         eps: float = None,
+        upsampfac: int = 2,
```
This should not be `int` since it can be 1.25. In fact, since it can only be either 1.25 or 2, I would give it the type `Literal[1.25, 2]`

steven-murray replied:
I also don't like `upsampfac` as a name, sorry! How about just `upsample_factor`?

tyler-a-cox replied:
No worries! I was just sticking to the naming convention used in `finufft` I've changed this to `upsample_factor` for more clarity.

steven-murray commented on src/fftvis/core/simulate.py:
```diff
@@ -193,6 +204,11 @@ def _evaluate_vis_chunk(
             Number of threads to use.
         is_coplanar : bool
             Whether the array is coplanar.
+        basis_matrix : np.ndarray
```
Maybe we should add that if None is given, we will automatically attempt to find one that would speed up the calculation.

tyler-a-cox replied:
Added more description here.

steven-murray commented on src/fftvis/core/simulate.py:
```diff
@@ -193,6 +204,11 @@ def _evaluate_vis_chunk(
+        basis_matrix : np.ndarray
+            Lattice basis matrix used to grid baselines and sources for a type 1
+            non-uniform FFT.
+        type1_n_modes : int
+            Number of modes for type 1 non-uniform FFT.
```
To me as a user, I would not understand what this means -- modes of what? What is the relevant range of values?

tyler-a-cox replied:
Added more description here to let users know that this will be the size of the output transform before indexing.

steven-murray commented on src/fftvis/cpu/cpu_simulate.py:
```diff
@@ -201,16 +203,12 @@ def simulate(
 
             # Assume the array is coplanar for gridded coordinates
             is_coplanar = True
+            rotation_matrix = np.eye(3, dtype=real_dtype)
```
Is this a stray reference to `rotation_matrix`?

tyler-a-cox replied:
No, this is just a placeholder for the rotation matrix that handles rotation of the array that has a slope along the xy and z-axes. I could also just change this so that the rotation matrix is an optional parameter, but we were already storing it as a shared array in `ray` so I just decided to keep it.

steven-murray commented on src/fftvis/cpu/cpu_simulate.py:
```diff
@@ -437,7 +437,14 @@ def _evaluate_vis_chunk(
 
                 # Rotate source coordinates with rotation matrix.
-                inplace_rot(rotation_matrix, topo)
+                if not np.allclose(rotation_matrix, np.eye(3)):
```
This seems like a small time-waste to me -- we should just have a boolean telling us if `rotation_matrix` is identity

tyler-a-cox replied:
Added an extra parameter which performs this check once outside the loop.

steven-murray commented on src/fftvis/utils.py:
```diff
@@ -32,9 +31,9 @@ def _use_gpu():
 # Import implementation-specific functions
 if _use_gpu():
-    from .gpu.gpu_utils import inplace_rot
+    from .gpu.utils import inplace_rot
```
I don't think this is the right way to do this -- this will set `inplace_rot` to either CPU or GPU for an entire session. This might not be what the user wants. We should instead use the correct `inplace_rot` whenever we need it, within a function.

tyler-a-cox replied:
I'm now just getting the CPU or GPU function directly rather than doing this.

steven-murray commented on tests/test_cpu_simulate.py:
```diff
@@ -171,6 +192,79 @@ def test_simulate(
+        ants = {k: ants[k] for k in ants if rng.uniform(0, 1) > 0.25}
+        ants = {ki: ants[k] for ki, k in enumerate(ants)}
+
+    # Remove antennas from the array
```
I think this comment is wrong
```suggestion
```

steven-murray commented on tests/test_cpu_simulate.py:
```diff
@@ -171,6 +192,79 @@ def test_simulate(
+        }
+    if shear_array:
+        shear_matrix = np.array([[1, 0.5, 0], [0, 1, 0], [0, 0, 1]])
+        # Sheer the array
```
```suggestion
        # Shear the array
```

steven-murray commented on tests/test_cpu_simulate.py:
```diff
@@ -171,6 +192,79 @@ def test_simulate(
+        **params,
+    )
+
+    # Check shape of result when no baselines are specified
```
```suggestion
    # Check that results are the same
```

steven-murray commented on tests/test_cpu_simulate.py:
```diff
@@ -171,6 +192,79 @@ def test_simulate(
+        coord_method="CoordinateRotationERFA",
+        source_buffer=0.75,
+        **params,
+    )
```
Why don't we test type1 vs type3 instead of testing both against matvis?

tyler-a-cox replied:
Now testing type 1/3 directly.

tyler-a-cox committed:
address documentation and implemenation comments

tyler-a-cox committed:
fix typos

tyler-a-cox committed:
remove enforcement of cpu/gpu utils

tyler-a-cox committed:
remove fftvis.utils.inplace_rot to force user to choose cpu/gpu implementation

steven-murray reviewed (approved):
Thanks @tyler-a-cox this looks very good now. Maybe the only thing would be to add an FAQ or tutorial to the docs describing this option.

tyler-a-cox committed:
add new notebook comparing type 1/3
